src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,415 @@
+  I.is_homogeneous 𝒜 ↔ (I : set A) ⊆ ⋂ i, graded_algebra.proj 𝒜 i ⁻¹' ↑I :=
+subset_Inter_iff.symm
+
+lemma ideal.is_homogeneous.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S)) ↔
+    I = I.homogeneous_core 𝒜 :=
+(set.image_preimage.compose (submodule.gi _ _).gc).exists_eq_l _
+
+lemma mul_homogeneous_element_mem_of_mem
+  {I : ideal A} (r x : A) (hx₁ : is_homogeneous 𝒜 x) (hx₂ : x ∈ I) (j : ι) :
+  graded_algebra.proj 𝒜 j (r * x) ∈ I :=
+begin
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw [←graded_algebra.sum_support_decompose 𝒜 r, finset.sum_mul, linear_map.map_sum],
+  apply ideal.sum_mem,
+  intros k hk,
+  obtain ⟨i, hi⟩ := hx₁,
+  have mem₁ : (graded_algebra.decompose 𝒜 r k : A) * x ∈ 𝒜 (k + i) := graded_monoid.mul_mem
+    (submodule.coe_mem _) hi,
+  rw [graded_algebra.proj_apply, graded_algebra.decompose_of_mem 𝒜 mem₁,
+    coe_of_submodule_apply 𝒜, submodule.coe_mk],
+  split_ifs,
+  { exact I.mul_mem_left _ hx₂ },
+  { exact I.zero_mem },
+end
+
+lemma ideal.is_homogeneous_homogeneous_core : (I.homogeneous_core 𝒜).is_homogeneous 𝒜 :=
+begin
+  rintros i r hr,
+  rw [ideal.homogeneous_core, ideal.span, finsupp.span_eq_range_total] at hr,
+  rw linear_map.mem_range at hr,
+  obtain ⟨s, rfl⟩ := hr,
+  rw [←graded_algebra.proj_apply, finsupp.total_apply, finsupp.sum, linear_map.map_sum],
+  refine ideal.sum_mem _ _,
+  rintros z hz1,
+  rw [smul_eq_mul],
+  refine mul_homogeneous_element_mem_of_mem 𝒜 (s z) z _ _ i,
+  { rcases z with ⟨z, hz2⟩,
+    rw subtype.image_preimage_coe at hz2,
+    exact hz2.2, },
+  { exact ideal.subset_span z.2 },
+end
+
+variables {𝒜 I}
+
+lemma ideal.is_homogeneous.homogeneous_core_eq_self (h : I.is_homogeneous 𝒜) :
+  I = I.homogeneous_core 𝒜 :=
+begin
+  apply le_antisymm _ (I.homogeneous_core_le_ideal 𝒜),
+  intros x hx,
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw ←graded_algebra.sum_support_decompose 𝒜 x,
+  refine ideal.sum_mem _ _,
+  intros j hj,
+  apply ideal.subset_span,
+  exact ⟨⟨_, is_homogeneous_coe _⟩, h _ hx, rfl⟩,
+end
+
+variables (𝒜 I)
+
+lemma ideal.is_homogeneous.iff_eq :
+  I.is_homogeneous 𝒜 ↔ I = I.homogeneous_core 𝒜 :=
+⟨ λ hI, hI.homogeneous_core_eq_self,
+  λ hI, hI.symm ▸ I.is_homogeneous_homogeneous_core 𝒜 ⟩
+
+lemma ideal.is_homogeneous.iff_exists :
+  I.is_homogeneous 𝒜 ↔ ∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S) :=
+by rw [ideal.is_homogeneous.exists_iff_eq_span, ideal.is_homogeneous.iff_eq]
+
+end is_homogeneous_ideal_defs
+
+section operations
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A] [algebra R A]
+variables [decidable_eq ι] [add_comm_monoid ι]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variable (I : ideal A)
+
+/--For any `comm_ring R`, we collect the homogeneous ideals of `R` into a type.-/
+abbreviation homogeneous_ideal : Type* := { I : ideal A // I.is_homogeneous 𝒜 }
```

Comment:
I think this definition should come as early as possible. Before the definition of the homogeneous core.

Reply:
done